src/main/java/tech/picnic/jolo/Relation.java

```diff
@@ -64,10 +63,58 @@
     this.leftSetter = leftSetter;
     this.rightSetter = rightSetter;
     this.relationLoader = relationLoader.orElse(this::foreignKeyRelationLoader);
-    this.relationLoaderIsCustom = relationLoader.isPresent();
+    relationLoaderIsCustom = relationLoader.isPresent();
   }
 
-  /** Copies this relation, discarding any state. This method is used in a prototype pattern. */
+  @Override
+  public boolean equals(@Nullable Object o) {
+    if (this == o) {
+      return true;
+    }
+    if (o == null || getClass() != o.getClass()) {
+      return false;
+    }
+    Relation<?, ?> relation = (Relation<?, ?>) o;
+    return relationLoaderIsCustom == relation.relationLoaderIsCustom
+        && Objects.equals(left, relation.left)
+        && Objects.equals(right, relation.right)
+        && Objects.equals(leftKey, relation.leftKey)
+        && Objects.equals(rightKey, relation.rightKey)
+        && leftArity == relation.leftArity
+        && rightArity == relation.rightArity
+        && Objects.equals(leftSetter, relation.leftSetter)
+        && Objects.equals(rightSetter, relation.rightSetter)
+        && Objects.equals(relationLoader, relation.relationLoader);
```

Comment:
Function equivalence is generally undecidable so I'm not sure if these setters should be included in this `equals` check. WDYT?